psi4/driver/driver.py

```diff
@@ -1938,17 +1938,111 @@ def fchk(wfn, filename):
     -----
     * A description of the FCHK format is http://wild.life.nctu.edu.tw/~jsyu/compchem/g09/g09ur/f_formchk.htm
     * The allowed headers for methods are general and limited, i.e., "Total SCF|MP2|CI|CC Density",
-      so "CC" is always used for the post-HF case.
+      PSI4 will try to find the right one for the current calculation.
+    * Not all implementations in PSI4 are compatible with the FCHK writer. Currently available and tested are:
+      All SCF/DFT variants, DF-RHF-MP2 (dfmp2 module), CONV-RHF/UHF-CCSD (ccenergy module). No UHF-MP2 densities are available.
 
     :examples:
 
     >>> # [1] FCHK file for DFT calculation
     >>> E, wfn = energy('b3lyp', return_wfn=True)
     >>> fchk(wfn, 'mycalc.fchk')
 
+    >>> # [2] FCHK file for correlated densities
+    >>> E, wfn = gradient('ccsd', return_wfn=True)
+    >>> fchk(wfn, 'mycalc.fchk')
+
     """
+    #NOTE: This function should be overhauled once all modules set Ca/Cb, Da/Db in the same manner.
+
+    # QC_MODULE overview for FCHK
+    # OCC: (includes both occ/dfocc modules
+    #  wfn.D() = SCF
+    #  wfn.C() = SCF
+    #  refwfn.D() = SCF
+    # -> not allowed, graceful exit
+    # DCT:
+    #  wfn.D() = SCF
+    #  wnf.C() = DCT
+    #  refwfn.D() = SCF
+    # -> allowed, but FCHK entries are switched around
+    # CCENERGY
+    #  wfn.D() = CC
+    #  wfn.C() = SCF 
+    #  refwfn.D() = CC
+    # -> allowed, will recompute SCF density from wfn.C()
+    # DFMP2:
+    #  wfn.D() = MP2
+    #  wnf.C() = NO(MP2)
+    #  refwfn.D() = SCF 
+    # -> allowed, will take refwfn.D() for SCF density
+
+    not_working_correctly = ['OCC']
+    qc_module = core.get_global_option('QC_MODULE')
+    if (qc_module in not_working_correctly[:]):
+        core.print_out(f"QC_MODULE {qc_module} is currently not supported by the FCHK writer.")
+        return None
+
     fw = core.FCHKWriter(wfn)
+
+    # At this point we don't the the method name, so we try to search for it.
+    # idea: get the method from the variable matching closely the 'current energy'
+    varlist = core.variables()
+    current = varlist['CURRENT ENERGY']
+    del varlist['CURRENT ENERGY']
+
+    # delete problematic entries
+    delete_list = ['CURRENT ENERGY', 'CURRENT GRADIENT', 'CURRENT REFERENCE ENERGY']
+    for key in delete_list[:]:
+        if key in varlist:
+            del varlist[key]
+
+    # find closest match, skip arrays
+    for (key, val) in varlist.items():
+        if hasattr(val, '__len__'): continue
```

Comment:
It looks like all ENERGY variables are scalar currently, so `endswith` is a good idea. Best would be if one can check if the entry is a scalar, instead of excluding arrays, various psi4.core objects, etc.